orchagent/nhgorch.cpp

```diff
@@ -0,0 +1,1386 @@
+    }
+
+    /* Set switch's next hop group capacity. */
+    vector<FieldValueTuple> fvTuple;
+    fvTuple.emplace_back("MAX_NEXTHOP_GROUP_COUNT",
+                         to_string(m_maxNhgCount));
+    gSwitchOrch->set_switch_capability(fvTuple);
+
+    SWSS_LOG_NOTICE("Maximum number of ECMP groups supported is %d",
+                    m_maxNhgCount);
+}
+
+/*
+ * Purpose:     Perform the operations requested by APPL_DB users.
+ *
+ * Description: Iterate over the untreated operations list and resolve them.
+ *              The operations supported are SET and DEL.  If an operation
+ *              could not be resolved, it will either remain in the list, or be
+ *              removed, depending on the case.
+ *
+ * Params:      IN  consumer - The cosumer object.
+ *
+ * Returns:     Nothing.
+ */
+void NhgOrch::doTask(Consumer& consumer)
+{
+    SWSS_LOG_ENTER();
+
+    if (!gPortsOrch->allPortsReady())
+    {
+        return;
+    }
+
+    auto it = consumer.m_toSync.begin();
+
+    while (it != consumer.m_toSync.end())
+    {
+        KeyOpFieldsValuesTuple t = it->second;
+
+        string index = kfvKey(t);
+        string op = kfvOp(t);
+
+        SWSS_LOG_INFO("Next hop group table key %s, op %s",
+                        index.c_str(), op.c_str());
+
+        bool success;
+        const auto& nhg_it = m_syncdNextHopGroups.find(index);
+
+        if (op == SET_COMMAND)
+        {
+            string ips;
+            string aliases;
+            string weights;
+            string mpls_nhs;
+
+            /* Get group's next hop IPs and aliases */
+            for (auto i : kfvFieldsValues(t))
+            {
+                if (fvField(i) == "nexthop")
+                    ips = fvValue(i);
+
+                if (fvField(i) == "ifname")
+                    aliases = fvValue(i);
+
+                if (fvField(i) == "weight")
+                    weights = fvValue(i);
+
+                if (fvField(i) == "mpls_nh")
+                    mpls_nhs = fvValue(i);
+            }
+
+            /* Split ips and alaises strings into vectors of tokens. */
+            vector<string> ipv = tokenize(ips, ',');
+            vector<string> alsv = tokenize(aliases, ',');
+            vector<string> mpls_nhv = tokenize(mpls_nhs, ',');
+
+            /* Create the next hop group key. */
+            string nhg_str;
+
+            for (uint32_t i = 0; i < ipv.size(); i++)
+            {
+                if (i) nhg_str += NHG_DELIMITER;
+                if (!mpls_nhv.empty() && mpls_nhv[i] != "na")
+                {
+                    nhg_str += mpls_nhv[i] + LABELSTACK_DELIMITER;
+                }
+                nhg_str += ipv[i] + NH_DELIMITER + alsv[i];
+            }
+
+            NextHopGroupKey nhg_key = NextHopGroupKey(nhg_str, weights);
+
+            /* If the group does not exist, create one. */
+            if (nhg_it == m_syncdNextHopGroups.end())
+            {
+                SWSS_LOG_NOTICE("Adding next hop group %s with %s",
```

Comment:
Please keep NOTICE level for the one that you really intend to be logged. Kindly revisit the log levels in PR

Reply:
Resolved